Xabe.AutoUpdater/FtpProvider.cs

```diff
@@ -0,0 +1,470 @@
+        {
+            _appName = appName;
+            _ftpHosts = pHosts;
+            _ftpLogin = pLogin;
+            _ftpPassiveMode = pPassiveMode;
+            _ftpPassword = pPassword;
+            _ftpFolder = string.IsNullOrEmpty(pFolder) ? "/" : pFolder;
+        }
+
+        /// <inheritdoc />
+        public Task<List<string>> DownloadCurrentVersion()
+        {
+            var t = Task.Run(async () =>
+            {
+                var client = BuildClient();
+                var latestVersion = await GetLatestVersionNumber();
+                if (!string.IsNullOrEmpty(latestVersion))
+                {
+                    var remoteFile = string.Concat(_ftpFolder, $"{_appName}.{latestVersion}.zip");
+                    var tempFile = Path.GetTempFileName();
+                    if (File.Exists(tempFile))
+                        File.Delete(tempFile);
+                    client.Download(remoteFile, tempFile);
+                    if (File.Exists(tempFile))
+                    {
+                        var outputDir = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+                        ZipFile.ExtractToDirectory(tempFile, outputDir);
+                        return Directory.GetFiles(outputDir, "*", SearchOption.AllDirectories)
+                            .ToList();
+                    }
+                }
+                return null;
+            });
+            t.Wait();
+            return t;
+        }
+
+        /// <inheritdoc />
+        public Task<string> GetLatestVersionNumber()
+        {
+            var t = Task.Run(() =>
+            {
+                string version = "0.0.0.0";
+                var client = BuildClient();
+                var remoteDirList = client.DirectoryListSimple(_ftpFolder);
+                if (remoteDirList != null && remoteDirList.Any())
+                {
+                    var filteredItems = remoteDirList.ToList().Where(p => p.StartsWith(_appName)).ToList();
+                    if (filteredItems.Any())
+                    {
+                        var lastItem = filteredItems.OrderByDescending(p => p).FirstOrDefault();
+                        if (lastItem != null)
+                        {
+                            // File Fomat : appName.Major.Minor.Build.Revision.zip
+                            version = Path.GetFileNameWithoutExtension(lastItem.Replace(_appName + ".", ""));
+                        }
+                    }
+                }
+                return version;
+            });
+            t.Wait();
+            return t;
+        }
+
+        private FtpClient BuildClient()
+        {
+            return new FtpClient
+            {
+                Host = _ftpHosts.StartsWith("ftp:") ? _ftpHosts : $"ftp://{_ftpHosts}",
+                User = _ftpLogin,
+                Pass = _ftpPassword,
+                Passive = _ftpPassiveMode,
+                KeepAlive = true
+            };
+        }
+
+    }
+
+    #region Ftp Client
```

Comment:
Regions are passe.
Please remove regions and move FtpClient to other file.

Reply:
Maybe it's passe, but code is more lisible sometimes. Ok will remove it